### DISCUSSION TITLE
Animations indexing wrong?

### DISCUSSION THREAD

**Discussion by mugmuggy** [CorsixTH/CorsixTH#1789] [General Diagnostics]
I've been looking at some original TH animation stuff and the objects, and this made me revisit some of that older documentation and animview (which I can't get to compile at the moment).

Specifically in animview
```c
struct th_anim_t {
  uint16_t frame;
  uint16_t unknown;
};
```
and in corsix itself
```c
  // Animation information structure reinterpreted from Theme Hospital data.
  struct th_animation_properties {
    uint16_t first_frame;
    // It could be that frame is a uint32_t rather than a uint16_t, which
    // would resolve the following unknown (which seems to always be zero).
    uint16_t unknown;
  } CORSIX_TH_PACKED_FLAGS;
```

The corsix comment itself, may hold true, but appears to be purely for alignment, the original itself only uses the lower word after reading in the entire dword. The issue I can identify with the structure is that its at least 4 bytes too short.

Take this hex dump, 8 byte aligned
![image](https://user-images.githubusercontent.com/8009292/103385807-ae105800-4b47-11eb-97a4-75ac347b7e6c.png)
The 2nd and 3rd, 5th and 8th and 6th and 7th record pairs, are duplicates of the other, with what appears to be a 4 byte reference to the same animation. Using 0 indexing of these records, the populated index, appears to be referring to the index record of the previous instance of the same animation if it exists or is required to do so.

What the purpose is, I don't know yet, some references take this index as given (as a dword), others check its not zero and and set it to true/false. A quick thought would be that it's used to drive the flipping/mirroring, rather than the check of an opposite orientation and setting the animaton flag that corsix currently uses if the opposite doesn't exist.

Obviously correcting it would mean every animation reference would change. However I assume that this is why we get fairly poor navigation through animview (0, 1, 2, 6), as its removing the duplicate references., which would appear to always skip the dword part when encountered.  Whereas with the 8 byte alignment it would be 0, 1, 3, 4, 5 and we would only have 2593 total in the count. Also explains why all animations are essentially a multiple of 2.

Animation 154 (in the current view - frame 165) would appear to have been used as a placeholder for animations that were not included, or removed, as it appears 381 times, but doesn't refer to itself.

**Comment by Alberth289346** [comment]
I would assume Corsix just use the reverse engineered TH specs which we have mirrored in https://github.com/CorsixTH/theme-hospital-spec. Could be useful to compare notes.

**Reply by mugmuggy** [reply]
I have my own fork, https://github.com/mugmuggy/theme-hospital-spec where I have done some edits previously for the saves (and still a work in progress) but just took the animation stuff as gospel, and was the documentation I referred too.

I went and modified some of those values to zero instead of the index, and indeed they are incorrectly mirrored. For example doctors moving south face the east or moving west face the north and north and east movements are ok.

**Comment by Alberth289346** [comment]
I think it makes sense to include mirror flags etc as part of the data file rather than having that hard-coded in the program.

To reach open source graphics, I believe we should have a list of animations that exist, including a suitable name, flags, used layers, and a length.
Lua code should refer to animations using only a name, thus creating the option for new graphics to have a completely different structure (eg create 4 animation in different directions instead of 2 + mirroring, or create all animations as a single frame image instead of having child sprites using layers).
